### ISSUE TITLE
Missing diagnostic for "-(1<<31)"

### ISSUE THREAD

**Issue by efriedma-quic** [llvm/llvm-project#31643]
|  |  |
| --- | --- |
| Bugzilla Link | [32296](https://llvm.org/bz32296) |
| Version | unspecified |
| OS | Windows NT |
| CC | @nlewycky |

## Extended Description 
Testcase:

int a = -(1<<31);

This is UB; we should warn by default.  (clang does warn for "0-(1<<31)".)

**Comment by xgupta** [comment]
The fix should be in this function-
https://github.com/llvm/llvm-project/blob/c872bce2ae9d27dbade67b1ed5443f29f4017451/clang/lib/AST/ExprConstant.cpp#L2718-L2741

I tried a few things, but none didn't work out, one like this  -
```
diff --git a/clang/lib/AST/ExprConstant.cpp b/clang/lib/AST/ExprConstant.cpp
index 2a5b5c3520b4..c84dd4209155 100644
--- a/clang/lib/AST/ExprConstant.cpp
+++ b/clang/lib/AST/ExprConstant.cpp
@@ -2723,6 +2723,12 @@ static bool CheckedIntArithmetic(EvalInfo &Info, const Expr *E,
                                  const APSInt &LHS, const APSInt &RHS,
                                  unsigned BitWidth, Operation Op,
                                  APSInt &Result) {
+
+if (!LHS) {
+  APSInt zero(32, false);
+  APSInt LHS(zero);
+}
+
   if (LHS.isUnsigned()) {
     Result = Op(LHS, RHS);
     return true;
```
If you have any suggestions, please let me know.

**Comment by shafik** [comment]
PR: https://reviews.llvm.org/D142867
